### PR TITLE
Refactor macho relocs patching and convert to vfiles

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**DO NOT SQUASH**

 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This removes one of the few remaining `io.cache` uses in bin and fixes some issues with macho relocs along the way. Some code from elf has been extracted into `librz/bin/relocs_patch.c` with minor fixes, hence the changes there.

**Test plan**

run the tests

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
This broken test is due to incorrectly handled arm32 relocs in elf. Both before and after are wrong. I am waiting for #1699 before fixing this.
```
[XX] /Users/florian/dev/rizin/test/db/formats/elf/elf-relarm ELF: arm sh relocs
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'ir~?
s section..init_array
pd 10
' bins/elf/r2pay-arm32.so
-- stdout
--- expected
+++ actual
@@ -6,8 +6,8 @@
             0x000b58c8      .dword 0x00004570 ; reloc.target..datadiv_decode12875440098136793118 ; sym..datadiv_decode12875440098136793118; RELOC 32 .datadiv_decode12875440098136793118 @ 0x00004570
             ;-- .datadiv_decode11893604691534341392:
             0x000b58cc      .dword 0x00098028 ; reloc.target..datadiv_decode11893604691534341392 ; sym..datadiv_decode11893604691534341392; RELOC 32 .datadiv_decode11893604691534341392 @ 0x00098028
-            0x000b58d0      .dword 0x001437fc                          ; RELOC 32
-            0x000b58d4      .dword 0x001437fe                          ; RELOC 32
+            0x000b58d0      .dword 0x00000000                          ; RELOC 32
+            0x000b58d4      .dword 0x00000000                          ; RELOC 32
             ;-- .datadiv_decode668687990735969933:
             0x000b58d8      .dword 0x00098b44 ; reloc.target..datadiv_decode668687990735969933 ; sym..datadiv_decode668687990735969933; RELOC 32 .datadiv_decode668687990735969933 @ 0x00098b44
             ;-- .datadiv_decode16868916699024037624:

-- stderr
Cannot determine entrypoint, using 0x00004524.

```

File: librz/bin/format/mach0/mach0.c [hunk 1]
```diff
@@ -8,12 +8,14 @@
 #include "mach0.h"
 #include <rz_msg_digest.h>
 
+#include "mach0_utils.inc"
+
 // TODO: deprecate bprintf and use Eprintf (bin->self)
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c]: Maybe change the comment to `Deprecate bprintf and eprintf and use RZ_LOG_*() functions instead`?

File: librz/bin/format/mach0/mach0.c [hunk 2]
```diff
@@ -36,18 +38,6 @@ typedef struct {
 // USE THIS: int ws = bf->o->info->big_endian;
 #define mach0_endian 1
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c:39]: Can you kill it too?

File: librz/bin/format/mach0/mach0.c [hunk 3]
```diff
@@ -2559,7 +2593,7 @@ static int inSymtab(HtPP *hash, const char *name, ut64 addr) {
 	return false;
 }
 
-static char *get_name(struct MACH0_(obj_t) * mo, ut32 stridx, bool filter) {
+RZ_API char *MACH0_(get_name)(struct MACH0_(obj_t) * mo, ut32 stridx, bool filter) {
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c]: Doxygen then? Also please mark the ownership of the returned string.

File: librz/bin/format/mach0/mach0_relocs.c [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+	size_t wordsize = get_word_size(bin);
+	if (bin->dyld_info) {
+		ut8 *opcodes, rel_type = 0;
+		size_t bind_size, lazy_size, weak_size;
+
+#define CASE(T) \
+	case ((T) / 8): rel_type = RZ_BIN_RELOC_##T; break
+		switch (wordsize) {
+			CASE(8);
+			CASE(16);
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0_relocs.c:137]: There would be missing `CASE(24)` after https://github.com/rizinorg/rizin/pull/1699 is finished.

File: librz/bin/format/mach0/mach0_relocs.c [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+			}
+			reloc->offset = sym == -1 ? 0 : bin->sects[i].offset + sym * wordsize;
+			reloc->addr = sym == -1 ? 0 : bin->sects[i].addr + sym * wordsize;
+			return true;
+		}
+	}
+	return false;
+}
+
+RZ_BORROW RzSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) * bin) {
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0_relocs.c:120]: This function is HUGE... What about splitting it up while you are on it?

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall the change is quite good.

File: librz/bin/format/mach0/mach0.c (hunk 2 above)
Comment (thestr4ng3r) [reply librz/bin/format/mach0/mach0.c:39]: Not yet, this needs further testing.

File: librz/bin/format/mach0/mach0_relocs.c (hunk 4 above)
Comment (thestr4ng3r) [reply librz/bin/format/mach0/mach0_relocs.c:137]: I don't think the wordsize can be 3.

File: librz/bin/format/mach0/mach0_relocs.c (hunk 5 above)
Comment (thestr4ng3r) [reply librz/bin/format/mach0/mach0_relocs.c:120]: I think it is too fragile. I would rather do that in another pr.

File: librz/bin/format/mach0/mach0_relocs.c (hunk 4 above)
Comment (XVilka) [reply librz/bin/format/mach0/mach0_relocs.c:137]: It's the case in some PowerPC relocs afaik, but @wargio is the expert in this field.

File: librz/bin/format/mach0/mach0.c [hunk 6]
```diff
@@ -2234,10 +2222,42 @@ RzList *MACH0_(get_virtual_files)(RzBinFile *bf) {
 		vf->name = strdup(MACH0_VFILE_NAME_REBASED_STRIPPED);
 		rz_list_push(ret, vf);
 	}
+
+	// relocs
+	MACH0_(patch_relocs)
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c]: Weird indentation.

File: librz/bin/format/mach0/mach0.c [hunk 7]
```diff
@@ -2272,6 +2292,20 @@ RzList *MACH0_(get_maps)(RzBinFile *bf) {
 	return ret;
 }
 
+RzList *MACH0_(get_maps)(RzBinFile *bf) {
+	RzList *ret = MACH0_(get_maps_unpatched)(bf);
+	if (!ret) {
+		return NULL;
+	}
+	struct MACH0_(obj_t) *obj = bf->o->bin_obj;
+	MACH0_(patch_relocs)
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0.c]: Weird indentation

File: librz/bin/format/mach0/mach0_relocs.c [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+	case CPU_TYPE_X86_64: {
+		switch (reloc->type) {
+		case X86_64_RELOC_UNSIGNED:
+			break;
+		case X86_64_RELOC_BRANCH:
+			pc -= 1;
+			ins_len = 5;
+			break;
+		default:
+			eprintf("Warning: unsupported reloc type for X86_64 (%d), please file a bug.\n", reloc->type);
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0_relocs.c]: RZ_LOG_WARN (remove that "Warning: ")

File: librz/bin/format/mach0/mach0_relocs.c [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+	}
+	case CPU_TYPE_ARM64:
+	case CPU_TYPE_ARM64_32:
+		pc = reloc->addr & ~3;
+		ins_len = 4;
+		break;
+	case CPU_TYPE_ARM:
+		break;
+	default:
+		eprintf("Warning: unsupported architecture for patching relocs, please file a bug. %s\n", MACH0_(get_cputype_from_hdr)(&bin->hdr));
```
Comment (XVilka) [inline librz/bin/format/mach0/mach0_relocs.c]: RZ_LOG_WARN

File: librz/bin/relocs_patch.c [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	return max + rz_num_align_delta(max, target_sz);
+}
+
+struct rz_bin_reloc_target_builder {
+	ut64 target_size; ///< size per target
+	HtUU *targets_by_sym; ///< to create only a single target per symbol, if there are more relocs for a single symbol
+	ut64 next_target;
+} /* RzBinRelocTargetBuilder */;
+
+RZ_API RzBinRelocTargetBuilder *rz_bin_reloc_target_builder_new(ut64 target_size, ut64 target_base) {
```
Comment (XVilka) [inline librz/bin/relocs_patch.c:35]: Please add Doxygen to this and the following functions.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/bin/format/mach0/mach0.c (hunk 6 above)
Comment (thestr4ng3r) [reply librz/bin/format/mach0/mach0.c]: clang-format enforces this unfortunately. I can turn it off there though.

**Review by XVilka — APPROVED** [review]

**Pull request merged**